source/fuzz/fuzzer_pass_add_parameters.cpp

```diff
@@ -70,14 +87,62 @@ void FuzzerPassAddParameters::Apply() {
         GetFuzzerContext()->GetRandomNumberOfNewParameters(
             GetNumberOfParameters(function));
     for (uint32_t i = 0; i < num_new_parameters; ++i) {
-      ApplyTransformation(TransformationAddParameter(
-          function.result_id(), GetFuzzerContext()->GetFreshId(),
-          // We mark the constant as irrelevant so that we can replace it with a
-          // more interesting value later.
-          FindOrCreateZeroConstant(
-              type_candidates[GetFuzzerContext()->RandomIndex(type_candidates)],
-              true),
-          GetFuzzerContext()->GetFreshId()));
+      uint32_t current_type_id =
+          type_candidates[GetFuzzerContext()->RandomIndex(type_candidates)];
+      auto current_type =
+          GetIRContext()->get_type_mgr()->GetType(current_type_id);
+      auto current_instr =
+          GetIRContext()->get_def_use_mgr()->GetDef(current_type_id);
+
+      if (current_type->kind() == opt::analysis::Type::kPointer) {
+        auto current_pointee_type_id =
+            fuzzerutil::GetPointeeTypeIdFromPointerType(GetIRContext(),
+                                                        current_type_id);
+        auto storage_class =
+            fuzzerutil::GetStorageClassFromPointerType(current_instr);
+
+        // Look for available variables that have the type |current_type|.
+        std::vector<uint32_t> available_variable_ids;
+        GetIRContext()->module()->ForEachInst(
+            [this, &available_variable_ids,
+             current_type_id](opt::Instruction* instruction) {
+              if (instruction->opcode() != SpvOpVariable) {
+                return;
+              }
+              if (instruction->type_id() != current_type_id) {
+                return;
+              }
+              available_variable_ids.push_back(instruction->result_id());
+            });
+        uint32_t initializer_id =
+            FindOrCreateZeroConstant(current_pointee_type_id, true);
+
+        // If there are no such variables, then create one. The value is
+        // irrelevant.
+        if (available_variable_ids.empty()) {
+          if (storage_class == SpvStorageClassPrivate) {
+            ApplyTransformation(TransformationAddGlobalVariable(
+                GetFuzzerContext()->GetFreshId(), current_type_id,
+                SpvStorageClassPrivate, initializer_id, true));
+          } else if (storage_class == SpvStorageClassFunction) {
+            ApplyTransformation(TransformationAddLocalVariable(
+                GetFuzzerContext()->GetFreshId(), current_type_id,
+                function.result_id(), initializer_id, true));
```

Comment:
This will create a local `OpVariable` instruction in the `function`. This is somewhat tricky.

The initial idea is that the created variable will be used as an operand to `OpFunctionCall` instructions. These instructions can belong to different functions - we'll call them callers - (since we can call any function from any other as long as there are no recursion - SPIR-V doesn't support that). Thus, this piece of code should create a local variable in every caller of `function.result_id()`. This is somewhat tricky because some callers might already have a suitable local variable.

I propose to refactor this fuzzer pass as follows:
1. Check the storage class of the variable. If it's `Private` or `Workgroup`, either find a global variable (`IRContext` must have a method that returns only global instructions) or create a new one (as you do above). Otherwise...
2. Iterate over all callers of `function.result_id()` (see `fuzzerutil::GetCallers`) and for each one either find a suitable local variable or create one.

I would also suggest to change the protobuf file to pass a repeated field of `UInt32Pair` (see `fuzzerutil::Repeated*ToMap` and `fuzzerutil::MapToRepeated*`). This field would represent a map from the result id of the caller to the result id of the local variable in that caller.

You would also need to adjust `IsApplicable` method in the transformation class to check that each caller has a local variable (if the added parameter is indeed a pointer with `Function` storage class). Hope this makes sense.

Reply:
I was thinking about the adding a map to a protobuf file and for me it seems unneccessary. I came up with two reasons:

1. It would mean refactoring the code in cases where AddParameter doesn't add a pointer with Function storage class.

2. In the case where AddParameter is about to add the pointer with Function storage class, the information whether the transformation is applicable or not is already accessible, since we pass function_id. And as you said, we could iterate over all callers and check if there is a local variable in each calling function. Even if we pass the map in the protobuf we still have to _recreate_ it in the IsApplicable method by iterating over all callers to check whether the resulting two maps are equal. For example, even if we have checked that each key in the passed protobuf map refers to a valid local variable, we might have missed some callers. The actual check is in iterating over the callers in IsApplicable method.

Reply:
Yes, you are right. It is indeed might be better to simply check in the transformation that the required parameter is available and in the fuzzer pass to ensure that the parameter is available. In this case, I would've created two functions in the fuzzerutil: `MaybeGetLocalVariable` and `MaybeGetGlobalVariable` (or some equivalent of these) that return result id of the local/global variable or 0 if no such variable exists.

You could then pass the type of the newly created parameter to the transformation and check that the variable (or the constant) of the required type exists.

There is one caveat, though. Currently, we do the opposite: we pass the value of the parameter and infer parameter's type out of it. You will need to change that.

Reply:
The function-creation part is up to you, though.